tensorflow/python/ops/ragged/ragged_embedding_ops.py

```diff
@@ -0,0 +1,242 @@
+      raise TypeError(f"sp_ids must be of the same type as sp_weights.")
+    sp_ids.values.get_shape().assert_is_compatible_with(
+        sp_weights.values.get_shape())
+    sp_ids.get_shape().assert_is_compatible_with(
+        sp_weights.get_shape())
+    # TODO(yleon): Add enhanced node assertions to verify that sp_ids and
+    # sp_weights have equal indices and shapes.
+
+  with ops.name_scope(name, "embedding_lookup_sparse",
+                      params + [sp_ids]) as name:
+
+    segment_ids = sp_ids.value_rowids()
+    ids = sp_ids.flat_values
+
+    if len(params) == 1 and max_norm is None and allow_dense_grads:
+      idx = ids
+      embeddings = params[0]
+    else:
+      ids, idx = array_ops.unique(ids)
+      embeddings = embedding_ops.embedding_lookup(
+          params, ids, partition_strategy=partition_strategy, max_norm=max_norm)
+
+    return embedding_ops.embedding_lookup_sparse_impl(embeddings, segment_ids,
+                                                      sp_weights, idx, combiner,
+                                                      ignore_weights, name)
+
+
+@dispatch.dispatch_for_api(embedding_ops.safe_embedding_lookup_sparse)
+def safe_embedding_lookup_sparse(embedding_weights,
+                                 sparse_ids: ragged_tensor.Ragged,
+                                 sparse_weights=None,
+                                 combiner="mean",
+                                 default_id=None,
+                                 name=None,
+                                 partition_strategy="div",
+                                 max_norm=None,
+                                 allow_dense_grads=False):
+  if embedding_weights is None:
+    raise ValueError(f"Missing embedding_weights {embedding_weights}.")
+  if isinstance(embedding_weights, variables.PartitionedVariable):
+    embedding_weights = list(embedding_weights)  # get underlying Variables.
+  if not isinstance(embedding_weights, list):
+    embedding_weights = [embedding_weights]
+  if len(embedding_weights) < 1:
+    raise ValueError(f"Missing embedding_weights {embedding_weights}.")
+
+  dtype = sparse_weights.dtype if sparse_weights is not None else None
+  embedding_weights = [
+      w if (isinstance(w, resource_variable_ops.ResourceVariable)
+            and dtype in (None, w.dtype))
+      else ops.convert_to_tensor(w, dtype=dtype)
+      for w in embedding_weights
+  ]
+
+  with ops.name_scope(name, "embedding_lookup", embedding_weights +
+                      [sparse_ids, sparse_weights]) as scope:
+
+    # Prune invalid ids and weights.
+    sparse_ids, sparse_weights = _prune_invalid_ids_ragged(sparse_ids,
+                                                            sparse_weights)
+    if combiner != "sum":
+      sparse_ids, sparse_weights = _prune_invalid_weights_ragged(
+          sparse_ids, sparse_weights)
+    sparse_ids, is_row_empty = ragged_array_ops.fill_empty_rows(
+      sparse_ids, default_id or 0)
```

Comment:
Curious, do you know which step it will fail if we remove the fill_empty_rows?

Reply:
The fill_empty_rows functionality (and the pruning of invalid entries) differentiates safe_embedding_lookup_sparse from embedding_lookup_sparse. Therefore, we can't remove it here.

Reply:
Yeah, it makes sense to keep it. I was just wondering why it is needed. 
I guess is only necessary to fill default_ids.